minetest/minetest
# Refactor builtin HUD

## Goal of the PR
- Refactor the builtin code for HUD elements, such that it can handle other HUD elements besides statusbars.
- It is needed for #14071 #14273 #14321 and potential future elements, for example for the chat or infotext.
- ~~(The statusbars behave the same as in the old version, the only difference is that `hud_replace_builtin` does not default `definition.item` anymore, but this is not documented in`lua_api.md` so it should be fine. However, I can re-add it if someone thinks it is needed.)~~ readded

## Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
2.2 Internal code refactoring

## To do

This PR is Ready for Review.

## How to test

- Start devtest, jump into water to test the breathbar, and hurt yourself to test the healthbar.
- Read the code notice the refactoring.
- Test `hud_replace_builtin`
```lua
minetest.hud_replace_builtin("breath", {
	type = "statbar",
	position = {x = 0.5, y = 0.5},
	text = "bubble.png",
	text2 = "bubble_gone.png",
	number = core.PLAYER_MAX_BREATH_DEFAULT * 2,
	item = core.PLAYER_MAX_BREATH_DEFAULT * 2,
	direction = 0,
	size = {x = 24, y = 24},
	offset = {x = 25, y= -(48 + 24 + 16)}
})
```
- Test flags
```lua
minetest.register_chatcommand("hudtoggleflag", {
	description = "Toggles a hud flag.",
	params = "<flag>",
	func = function(name, params)
		local player = minetest.get_player_by_name(name)
		if not player then
			return false, "No player."
		end

		local flags = player:hud_get_flags()
		if flags[params] == nil then
			return false, "Unknown hud flag."
		end

		flags[params] = not flags[params]
		player:hud_set_flags(flags)
		return true, "Flag \"".. params .."\" set to ".. tostring(flags[params]) .. "."
	end
})
```
`/hudtoggleflag breathbar` `/hudtoggleflag healthbar`
- Also test other builtin elements in case new ones get added by other PRs.
